text/0000-hash-simplification.md

```diff
@@ -0,0 +1,405 @@
+**not** reproducible outside of Rust. For this reason, APIs for reproducible
+hashes to be verified elsewhere will explicitly not be considered in the design
+for `std::hash`. It is expected that an external crate may wish to provide a
+trait for these hashing algorithms and it would not be bounded by
+`std::hash::Hash`, but instead perhaps a "byte container" of some form.
+
+# Detailed design
+
+This RFC considers two possible designs as a replacement of today's `std::hash`
+API. One is a "minor refactoring" of the current API while the
+other is a much more radical change towards being conservative. This section
+will propose the more radical change and the other may be found in the
+[Alternatives](#alternatives) section.
+
+## API
+
+The new API of `std::hash` will be:
+
+```rust
+trait Hash {
+    fn hash(&self) -> usize;
+}
+
+fn combine(a: usize, b: usize) -> usize;
+```
+
+The `Writer`, `Hasher`, and `SipHasher` structures/traits will all be removed
+from `std::hash`. This definition is more or less the Rust equivalent of the
+Java/C++ hashing infrastructure. This API is a vast simplification of what
+exists today and allows implementations of `Hash` as well as consumers of `Hash`
+to quite ergonomically work with hash values as well as hashable objects.
+
+> **Note**: The choice of `usize` instead of `u64` reflects [C++'s
+> choice][cpp-hash] here as well, but it is quite easy to use one instead of
+> the other.
+
+## Hashing algorithm
+
+With this definition of `Hash`, each type must pre-ordain a particular hash
+algorithm that it implements. Using an alternate algorithm would require a
+separate newtype wrapper.
+
+Most implementations will still use `#[derive(Hash)]` which will leverage
+`hash::combine` to combine the hash values of aggregate fields. Manual
+implementations which only want to hash a select number of fields would look
+like:
+
+```rust
+impl Hash for MyType {
+    fn hash(&self) -> usize {
+        // ignore field2
+        (&self.field1, &self.field3).hash()
+    }
+}
+```
+
+A possible implementation of combine can be found [in the boost source
+code][boost-combine].
+
+[boost-combine]: https://github.com/boostorg/functional/blob/master/include/boost/functional/hash/hash.hpp#L209-L213
+
+## `HashMap` and DoS protection
+
+Currently one of the features of the standard library's `HashMap` implementation
+is that it by default provides DoS protection through two measures:
+
+1. A strong hashing algorithm, SipHash 2-4, is used which is fairly difficult to
+   find collisions with.
+2. The SipHash algorithm is randomly seeded for each instance of `HashMap`. The
+   algorithm is seeded with a 128-bit key.
+
+These two measures ensure that each `HashMap` is randomly ordered, even if the
+same keys are inserted in the same order. As a result, it is quite difficult to
+mount a DoS attack against a `HashMap` as it is difficult to predict what
+collisions will happen.
```

Comment:
Note that it is not strictly necessary to randomly seed each instance in order to avoid hashDoS. All of the instances of hashDoS can share the same random seed.